engineering/architecture/numbers.md

```diff
@@ -0,0 +1,106 @@
+
+## Implementation strategies
+
+### API
+
+- The API transmits numbers as strings because most JSON parsers are not capable of accurately parsing high precision numbers or special Postgres numbers stored plainly in JSON. The format for such strings is described below.
+
+### Canonical stringified number format
+
+Throughout our stack, we may want to represent numbers as strings. When doing so, we follow a format very similar to numbers in JSON.
+
+- The decimal separator is an ASCII dot.
+- The minus sign is an ASCII hyphen.
+- The exponent separator is an uppercase "E".
+- Grouping separators are disallowed.
+- The strings `"NaN"`, `"-Infinity"`, and `"Infinity"` are used to represent special Postgres numbers. Note that this is one way in which our stringified number format differs from JSON -- raw values like `NaN` are _not_ valid inside JSON.
+- The string `"NULL"` should _not_ be used. In JSON, the _value_ `null` should be used instead.
+- Examples:
+
+    - `"NaN"`
+    - `"-Infinity"`
+    - `"Infinity"`
+    - `"0"`
+    - `"1234.5"`
+    - `"99999999999999999999999999999999"`
+    - `"-1.2E-15"`
+
+
+### Locale
+
+- To satisfy Goal D, we infer the user's desired locale from their _browser_. The back-end is not aware of the user's locale unless the front end passes data about its in the API request.
+
+- Mathesar is limited to only displaying and accepting entry for numbers using the latin numbering system. This means sacrificing Goal A for the locales  "bn", "fa", and "mr" (Bengali, Persian, Marathi) which use non-latin numbering systems by default.
+
+- The minus sign is always displayed using an ASCII hyphen, regardless of locale (even though some locales like 'li' use the Unicode minus sign). The Unicode minus sign (and similar characters) will be accepted in all locales while _entering_ numbers though.
+
+- For display of scientific notation, the exponent separator is always displayed using the ASCII "E" regardless of locale even though some locales use other exponent separators. For input, other strings are accepted too, such as "e" (lowercase) and "Е" (U+0415) and variations of "x10^".
+
+### Parsing
+
+- The front end parses **user input**, by performing string transformations to normalize the input into canonical stringified form. For example, all characters matching the current locale's decimal separator are converted to ASCII dots.
+
+- The front end parses **canonical stringified numbers** (as produced above, or received from the API) by attempting to produce a `number` or `BigInt` which can be re-formatted exactly to match the input (without loss of precision).
+
+- The front end will **fail** to parse canonical stringified numbers like `"99999999999999.99"` (which cannot be accurately represented by any of JavaScript's built-in data types) because they will fail to be reformatted without loss of precision.
```

Comment:
Is there a reason why we need to parse it to a number type on the frontend, after we normalize the string? What advantage does having a number type provide us, compared to directly using the normalized string? 

I'm only currently considering the usecase of entering numbers on the table cell, and not regular usage of the component.

Reply:
Right -- yeah good question. I considered the idea of keeping numbers only as strings on the front end. Broadly there are two reasons:

1. Formatting a cell's numerical value for display. There are a variety of smaller goals here.

    - Likely the most important goal is to format the decimal separator in accordance with the user's locale (dot or comma). That's easy to do with strings, so if that's all we care about, then yeah, we can stick with strings.
    - Adding grouping separators is _nice_, but not crucial. Anecdotally, at my previous job I pushed a feature which added grouping separators as users typed into a field where they were frequently entering larger numbers -- and we got a lot of positive feedback about it immediately. Even if the separators are applied after the user finishes typing, I think they offer a significant UX improvement in some situations. If we stick exclusively to strings, it's actually quite hard to add the grouping separators in a locale-aware way. The `Intl` API takes care of it for us, but only if we have a `number` or `BigInt` -- so that requires parsing.
    - Similar to above, `Intl` offers a bunch of other [formatting options](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/NumberFormat) that we'll likely want to utilize at some point in the future, like `style: 'percent'`, `style: 'currency'`, `minimumFractionDigits`, and `maximumFractionDigits`. That stuff is trivial to implement when using `Intl`, but harder to do with raw strings.

1. Accepting a user-entered number in other contexts.

    This is the "regular usage of the component", as you say. This concern is not actually a pressing need for the features we're currently building, but more of a concern about building a robust and intuitive component library (which lowers the priority here, but still warrants consideration as we make these low-level architectural decisions). For editing a cell, I doubt we'll ever need access to the _numerical value_ of the cell on the front end (beyond the need to format the number for display). But we likely will need access to numerical values of user input on the front end in the future. For example, when/if we allow the user to configure the minimum number of fraction digits, I'd want to take that user input and feed it back into `Intl.NumberFormat` to display a formatted preview of a sample number. I think we'd also want client-side validation to ensure that `minimumFractionDigits` is within `0` and `20`, as required by `Intl.NumberFormat`.

Parsing the number is more work, but not _that_ much more, especially given the code I already have in-place in #1155. To finish out that PR, my plan is as follows:

- Move the majority of the code in the `NumberFormatter` class into a class called `UniversalNumberFormatter` which, instead of implementing `InputFormatter<number>`, implements `InputFormatter<UniversalNumberParseValue>` where `UniversalNumberParseValue` is as follows:

    ```ts
    export type UniversalParseResultType = 'number' | 'bigFloat' | 'bigInt';
    interface BaseParseValue {
      /**
      * A canonical string representation of the number, like you'd see in a JSON
      * string. No grouping separators. ASCII only. Dot for decimal separator.
      */
      string: string;
    }
    export interface SimpleNumberParseValue extends BaseParseValue {
      type: 'number';
      number: number;
    }
    export interface BigIntNumberParseValue extends BaseParseValue {
      type: 'bigInt';
      bigInt: bigint;
    }
    export interface BigFloatParseValue extends BaseParseValue {
      type: 'bigFloat';
    }
    export type UniversalNumberParseValue =
      | SimpleNumberParseValue
      | BigIntNumberParseValue
      | BigFloatParseValue;
    ```

- Then, by composing `UniversalNumberFormatter`, I'll build a very simple `NumberFormatter` class which implements `InputFormatter<number>` and a very simple `StringifiedNumberFormatter` which implements `InputFormatter<string>`. I'll use those classes to build very simple components `NumberInput.svelte` and `StringNumberInput.svelte`. For cell editing, we'd use `StringNumberInput` which handles high-precision floats. For other use-cases in the future where the front end needs access to numerical values from user input, we'd use `NumberInput` which prevents the user from entering values which cannot be accurately represented by `number`. Does that make sense?

Reply:
Okay, sounds good to me.

Reply:
@seancolsen I think your comment here would be a good addition to the spec for future reference in case someone has the same question as @pavish.

Reply:
@kgodey 

> your comment here would be a good addition to the spec

Fixed in 8d79053876b7deb1b86951f93e66414db099686d